modules/ximgproc/src/deriche_filter.cpp

```diff
@@ -0,0 +1,410 @@
+
+    public:
+        ParallelGradientDericheXRows(Mat& imgSrc, Mat &d, double ald) :
+            img(imgSrc),
+            dst(d),
+            alphaDerive(ald),
+            verbose(false)
+        {}
+        void Verbose(bool b) { verbose = b; }
+        virtual void operator()(const Range& range) const
+        {
+            CV_Assert(img.depth()==CV_8UC1 || img.depth()==CV_8SC1 || img.depth()==CV_16SC1 || img.depth()==CV_16UC1);
+            CV_Assert(dst.depth()==CV_32FC1);
+            if (verbose)
+                std::cout << getThreadNum() << "# :Start from row " << range.start << " to " << range.end - 1 << " (" << range.end - range.start << " loops)" << std::endl;
+
+            switch (img.depth()) {
+            case CV_8U:
+                unsigned char c1;
+                HorizontalIIRFilter(img,dst,range, &c1,alphaDerive);
+            break;
+            case CV_8S:
+                char c2;
+                HorizontalIIRFilter(img, dst, range, &c2, alphaDerive);
+                break;
+            case CV_16U:
+                ushort c3;
+                HorizontalIIRFilter(img, dst, range, &c3, alphaDerive);
+                break;
+            case CV_16S:
+                short c4;
+                HorizontalIIRFilter(img, dst, range, &c4, alphaDerive);
+                break;
+                break;
+            default:
+                return;
+            }
+        };
+        ParallelGradientDericheXRows& operator=(const ParallelGradientDericheXRows &) {
+            return *this;
+        };
+    };
+
+    void GradientDericheY(InputArray _op, OutputArray _dst,double alphaDerive, double alphaMean)
+    {
+        Mat tmp(_op.size(), CV_32FC(_op.channels()));
+        _dst.create( _op.size(),CV_32FC(tmp.channels()) );
+        Mat opSrc = _op.getMat();
+        std::vector<Mat> planSrc;
+        split(opSrc, planSrc);
+        std::vector<Mat> planTmp;
+        split(tmp, planTmp);
+        std::vector<Mat> planDst;
+        split(_dst, planDst);
+        for (size_t i = 0; i < planSrc.size(); i++)
+        {
+            CV_Assert(planSrc[i].isContinuous() && planTmp[i].isContinuous() && planDst[i].isContinuous());
+            ParallelGradientDericheYCols x(planSrc[i], planTmp[i], alphaDerive);
+            parallel_for_(Range(0, opSrc.cols), x, getNumThreads());
+            ParallelGradientDericheYRows xr(planTmp[i], planDst[i], alphaMean);
+            parallel_for_(Range(0, opSrc.rows), xr, getNumThreads());
+
+        }
+        merge(planDst, _dst);
+    }
+
+    void GradientDericheX(InputArray _op, OutputArray _dst, double alphaDerive, double alphaMean)
+    {
+        Mat tmp(_op.size(), CV_32FC(_op.channels()));
+        _dst.create( _op.size(),CV_32FC(tmp.channels()) );
+        Mat opSrc = _op.getMat();
+        std::vector<Mat> planSrc;
+        split(opSrc, planSrc);
+        std::vector<Mat> planTmp;
+        split(tmp, planTmp);
+        std::vector<Mat> planDst;
+        split(_dst, planDst);
```

Comment:
I don't understand I use it at line https://github.com/LaurentBerger/opencv_contrib/blob/DericheFilter/modules/ximgproc/src/deriche_filter.cpp#L401